File: Class 12/birthdayparty.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

void dfs(int node, unordered_map<int, vector<int>>& adjlist, vector<bool>& visited, vector<int>& component) {
    visited[node - 1] = true;
    component.push_back(node);

    for (int neighbor : adjlist[node]) {
        if (!visited[neighbor - 1]) {
            dfs(neighbor, adjlist, visited, component);
        }
    }
}

int main() {
    int N, M, Q;
    cin >> N >> M >> Q;

    unordered_map<int, vector<int>> adjlist;

    for (int i = 1; i <= N; ++i) {
        adjlist[i] = vector<int>();
    }

    for (int i = 0; i < M; ++i) {
        int ai, bi;
        cin >> ai >> bi;
        adjlist[ai].push_back(bi);
        adjlist[bi].push_back(ai);
    }

    vector<bool> visited(N, false);
    unordered_map<int, int> components; 
    int component_id = 0;

    for (int i = 1; i <= N; ++i) {
        if (!visited[i - 1]) {
            vector<int> component;
            dfs(i, adjlist, visited, component);
            for (int node : component) {
                components[node] = component_id;
            }
            component_id += 1;
        }
    }

    for (int i = 0; i < Q; ++i) {
        int xj, yj;
        cin >> xj >> yj;
        if (components[xj] == components[yj]) {
            cout << "Y" << endl;
        } else {
            cout << "N" << endl;
        }
    }

    return 0;
}

File: Class 12/birthdayparty.py
def dfs(node, adjlist, visited, component):
    visited[node - 1] = True
    component.append(node)
    
    for neighbor in adjlist[node]:
        if not visited[neighbor - 1]:
            dfs(neighbor, adjlist, visited, component)


N, M, Q = map(int, input().split(" "))
adjlist = {}

for i in range(1, N + 1):
    adjlist[i] = []

for i in range(M):
    ai, bi = map(int, input().split(" "))
    adjlist[ai].append(bi)
    adjlist[bi].append(ai)

visited = [False] * N  
components = {}  
component_id = 0 

for i in range(1, N + 1):
    if not visited[i - 1]:
        component = []
        dfs(i, adjlist, visited, component)
        for node in component:
            components[node] = component_id
        component_id += 1


for i in range(Q):
    xj, yj = map(int, input().split(" "))
    if components[xj] == components[yj]:
        print("Y")
    else:
        print("N")
